### PR TITLE
Minor: Avoid recomputing compute_array_ndims in align_array_dimensions

### LINKED ISSUE #8024: Avoid recomputing compute_array_ndims in align_array_dimensions
### Is your feature request related to a problem or challenge?

Cleanup code

### Describe the solution you'd like

Have one compute_array_ndims and reuse it.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #8024.

## Rationale for this change

Main cleanup is to remove the duplicate compute of `compute array ndim`

### CODE REVIEW

File: datafusion/physical-expr/src/array_expressions.rs [hunk 1]
```diff
@@ -743,35 +743,39 @@ pub fn array_prepend(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+        let ndim = compute_array_ndims(Some(arg.to_owned()))?;
+        if let Some(ndim) = ndim {
+            args_ndim.push(ndim);
+        } else {
+            return internal_err!("args should not be empty");
+        }
+    }
+
+    let max_ndim = args_ndim.iter().max();
+    let max_ndim = if let Some(max_ndim) = max_ndim {
```
Comment (comphead) [inline datafusion/physical-expr/src/array_expressions.rs:757]: this check can be earlier, before the loop?

File: datafusion/physical-expr/src/array_expressions.rs [hunk 2]
```diff
@@ -1946,6 +1950,47 @@ mod tests {
...
+            Arc::new(ListArray::from_iter_primitive::<Int64Type, _, _>(vec![
+                Some(vec![Some(6), Some(7), Some(8)]),
+            ]));
+
+        let array2d_1 = Arc::new(wrap_into_list_array(array1d_1.clone())) as ArrayRef;
+        let array2d_2 = Arc::new(wrap_into_list_array(array1d_2.clone())) as ArrayRef;
+
+        let res =
+            align_array_dimensions(vec![array1d_1.to_owned(), array2d_2.to_owned()])
+                .expect("should not error");
```
Comment (comphead) [inline datafusion/physical-expr/src/array_expressions.rs:1970]: Please add more descriptive text in `expect`

**Comment by comphead** [comment]
what exactly issue the PR closes?

File: datafusion/physical-expr/src/array_expressions.rs (hunk 2 above)
Comment (jayzhan211) [reply datafusion/physical-expr/src/array_expressions.rs:1970]: Could I change them all to unwrap for test?

**Comment by jayzhan211** [comment]
> what exactly issue the PR closes?

Nope

File: datafusion/physical-expr/src/array_expressions.rs (hunk 2 above)
Comment (Weijun-H) [reply datafusion/physical-expr/src/array_expressions.rs]: I think we could only unwrap in tests based on this [influxdata/influxdb@main/docs/style_guide.md#dont-return-result-from-test-functions](https://github.com/influxdata/influxdb/blob/main/docs/style_guide.md?rgh-link-date=2023-10-26T16%3A42%3A50Z#dont-return-result-from-test-functions)

**Review by comphead — COMMENTED** [review]
Thanks @jayzhan211 I'll check this later today, please add the github issue what this ticket is supposed to close. Personally for me, I'm not really sure what exactly we improving or fixing, is it some issue requiring this fix? Please create a github issue with details and connect it to this PR

**Comment by jayzhan211** [comment]
> Thanks @jayzhan211 I'll check this later today, please add the github issue what this ticket is supposed to close. Personally for me, I'm not really sure what exactly we improving or fixing, is it some issue requiring this fix? Please create a github issue with details and connect it to this PR

I don't think there is such issue or any issue need to

> Thanks @jayzhan211 I'll check this later today, please add the github issue what this ticket is supposed to close. Personally for me, I'm not really sure what exactly we improving or fixing, is it some issue requiring this fix? Please create a github issue with details and connect it to this PR

This is just a minor improvement that remove the duplicated `compute_ndim_arrays`, I don't think we really need to create every issue for such minor improvement.

**Comment by comphead** [comment]
> > Thanks @jayzhan211 I'll check this later today, please add the github issue what this ticket is supposed to close. Personally for me, I'm not really sure what exactly we improving or fixing, is it some issue requiring this fix? Please create a github issue with details and connect it to this PR
> 
> This is just a minor improvement that removes the duplicated `compute_ndim_arrays`, I don't think we really need to create every issue for such a minor improvement.

That is true, for minors sometimes people avoid the originating github issue, which basically a little bit breaks the issue tracking. For the implementation, the its a nice idea to create a collection ndims along with separate max value and then zip the collection. However the complexity of method explodes with conditional statememts, I strongly believe it can be improved

**Review by Weijun-H — APPROVED** [review]
LGTM! Thank you @jayzhan211

**Review by comphead — APPROVED** [review]
lgtm, that looks much cleaner @jayzhan211

**Pull request merged**